sbncode/LArRecoProducer/CRUMBS_module.cc

```diff
@@ -0,0 +1,681 @@
+
+#include "nusimdata/SimulationBase/MCTruth.h"
+#include "nusimdata/SimulationBase/MCParticle.h"
+
+#include "lardataobj/RecoBase/Slice.h"
+#include "lardataobj/RecoBase/PFParticle.h"
+#include "lardataobj/RecoBase/Track.h"
+#include "lardataobj/RecoBase/Hit.h"
+#include "lardataobj/RecoBase/PFParticleMetadata.h"
+#include "lardataobj/AnalysisBase/T0.h"
+#include "lardataobj/AnalysisBase/Calorimetry.h"
+
+#include "lardata/DetectorInfoServices/DetectorClocksService.h"
+#include "larsim/Utils/TruthMatchUtils.h"
+
+#include "sbncode/GeometryTools/TPCGeoAlg.h"
+#include "sbnobj/Common/Reco/SimpleFlashMatchVars.h"
+#include "sbnobj/Common/Reco/StoppingChi2Fit.h"
+#include "sbncode/LArRecoProducer/TrackStoppingChi2Alg.h"
+#include "sbnobj/Common/Reco/CRUMBSResult.h"
+
+#include "TTree.h"
+#include "TMVA/Reader.h"
+
+class CRUMBS;
+
+namespace sbn {
+  class CRUMBS : public art::EDProducer {
+  public:
+    explicit CRUMBS(fhicl::ParameterSet const& p);
+    // The compiler-generated destructor is fine for non-base
+    // classes without bare pointers or other resource use.
+
+    // Plugins should not be copied or assigned.
+    CRUMBS(CRUMBS const&) = delete;
+    CRUMBS(CRUMBS&&) = delete;
+    CRUMBS& operator=(CRUMBS const&) = delete;
+    CRUMBS& operator=(CRUMBS&&) = delete;
+
+    // Required functions.
+    void produce(art::Event& e) override;
+
+    void ClearMaps();
+    void ResetVars();
+    void SetupMaps(art::Event const& e);
+
+    art::Ptr<recob::PFParticle> GetSlicePrimary(art::Event const& e, const art::Ptr<recob::Slice> slice, const art::ValidHandle<std::vector<recob::Slice> > handleSlices);
+
+    std::vector<art::Ptr<anab::T0> > GetCRTTrackT0s(art::Event const& e, const art::Ptr<recob::Slice> slice, art::ValidHandle<std::vector<recob::PFParticle> > handlePFPs,
+						    art::ValidHandle<std::vector<recob::Slice> > handleSlices);
+    std::vector<art::Ptr<anab::T0> > GetCRTHitT0s(art::Event const& e, const art::Ptr<recob::Slice> slice, art::ValidHandle<std::vector<recob::PFParticle> > handlePFPs,
+						  art::ValidHandle<std::vector<recob::Slice> > handleSlices);
+    float GetLongestTrackStoppingChi2Ratio(art::Event const& e, const art::Ptr<recob::Slice> slice, art::ValidHandle<std::vector<recob::PFParticle> > handlePFPs,
+					   art::ValidHandle<std::vector<recob::Slice> > handleSlices);
+
+    void FillCRTVars(std::vector<art::Ptr<anab::T0> > trackT0s, std::vector<art::Ptr<anab::T0> > hitT0s);
+    void FillPandoraNuScoreVars(std::map<std::string, float> propertiesMap);
+
+    std::vector<art::Ptr<recob::Hit> > GetAllSliceHits(art::Event const& e, const art::Ptr<recob::Slice> slice, art::ValidHandle<std::vector<recob::Slice> > handleSlices);
+    std::map<int,float> SlicePurity(art::Event const& e, std::vector<art::Ptr<recob::Hit> > sliceHits);
+    float SliceCompleteness(art::Event const& e, std::vector<art::Ptr<recob::Hit> > sliceHits, std::vector<art::Ptr<recob::Hit> > allHits, const int matchedGenID);
```

Comment:
I haven't checked in detail, but are all these `Ptr`s and `Handle`s really necessary? Could most of these arguments just be simple references?

Reply:
Is there a good reason for changing them? Clearly this is my inexperience showing but I've tended to pass ptrs and handles in other analyzers and producers I've written before?

Reply:
I guess mostly to reduce clutter, so you can see what actually matters. You only need a Ptr (say) if you want to look it up in an Assn, or other art-like thing. Normally I think it's supposed to be good practice to keep the stuff for dealing with the art interfaces, and the stuff for actually doing the physics, as separated as possible.

Reply:
This is another one where I'm going to opt for leaving it as is unless anyone feels its a merge-stopping case. I think the code is followable (sometimes more so) and seems to be relatively common place.